### ISSUE TITLE
Cannot use Catalog ImageBuilder for custom product image attribute

### ISSUE THREAD

**Issue by erfanimani** [magento/magento2#10161]
### Preconditions

1. Magento 2.1.7

### Steps to reproduce

1. Create product attribute with input type media image, and attribute code "side_image" for example
2. Create a new product, add a few images, and select the role of one of the images to be the "side_image" 
3. In our template we want to show the side image (using the block `\Magento\Catalog\Block\Product\View`). Therefore we're adding the following code:

```
<?php /** @var $block \Magento\Catalog\Block\Product\View */ ?>
<?php
    $product = $block->getProduct();

    $image = $block->getImage(
        $product,
        'product_page_image_medium',
        ['type' => 'side_image']
    );
?>

<img src="<?php echo $image->getImageUrl() ?>" />

```

### Expected result

1. Product side image shows up

### Actual result

1. Product base image shows up

A stack exchange question related to this bug can be found here: https://magento.stackexchange.com/questions/102498/extending-the-complextype-named-imagetype-with-a-custom-image-type

Essentially, without having to specify the third parameter, one would expect to create new image types by adding them into the view.xml, like this:
```
    $image = $block->getImage(
        $product,
        'my_custom_image_id_with_custom_product_image_type'
    );
```
.. however, that's not possible given XSD validation, and hence devs are trying to overwrite the core XSD to simply add their own image types.

The problem here, however, is that the third parameter gets discarded along the way, and that definitely seems like a bug to me.

On top of that, the fact that it's not possible to add image types definitely seems like a design flaw, and maybe the XSD shouldn't be this strict.

**Comment by erfanimani** [comment]
Actually, I found out that `attributes` is for custom HTML attributes on the `img` tag, not for image processing. The problem still stands - there is actually no way we can add our own custom image types.

**Comment by koenner01** [comment]
Definitely a major flaw in the system. We are allowed to create image attributes but can't use them...

**Comment by koenner01** [comment]
There might be some hope in this commit https://github.com/magento/magento2/commit/7fa3789b99955f0a4358fa43f43aff3cac7cee79 . It allows for the constructor to take in a path different from the default.

**Comment by koenner01** [comment]
Ok after experimenting with the commit I referenced in the previous post, I can confirm that it simplifies the whole need to plugin the system. After patching in the commit, you can set the argument directly in your di.xml now.

Example view.xml

```
<?xml version="1.0"?>
<view xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Config/etc/view.xsd">
    <media>
        <images module="Magento_Catalog">
            <image id="fetch_my_new_fancy_type" type="some_fancy_new_image_type">
                <width>600</width>
                <height>600</height>
            </image>
        </images>
    </media>
</view>
```
Example di.xml

```
<type name="Magento\Framework\Config\SchemaLocator">
    <arguments>
        <argument name="realPath" xsi:type="string">urn:magento:module:Vendor_MyModule:etc/view.xsd</argument>
    </arguments>
</type>
```

This is still not the _right_ fix for the problem.
If only the noNamespaceSchemaLocation file would be used correctly if we filled in our own view.xsd. That would solve all these workarounds..

**Comment by magento-engcom-team** [comment]
@erfanimani Thanks for reporting this issue. Please check http://devdocs.magento.com/guides/v2.1/frontend-dev-guide/themes/theme-images.html
Also You can try change Your code to something like:
```
<?php
    $product = $block->getProduct();
    $image = $block->getImage(
	$product,
	'side_image'
    );
?>
<img src="<?php echo $image->getImageUrl() ?>" />
```
And add to app/design/frontend/Magento/luma/etc/view.xml lines:
```
<image id="side_image" type="side_image">
	<width>700</width>
	<height>560</height>
</image>
```
Hope it helps. 
Please reopen this issue as needed.

**Comment by koenner01** [comment]
@magento-engcom-team, the documentation you are linking is either lacking or is incorrect.
For 'type' it states:

> The type of the images defined by the specified id. Allowed values:
> image - corresponds to the Base Image role in the Magento Admin
> small_image - corresponds to the Small Image role in the Magento Admin
> swatch_image - corresponds to the Swatch Image role in the Magento Admin
> swatch_thumb - corresponds to the Swatch Image role in the Magento Admin.
> thumbnail - corresponds to the Thumbnail Image role in the Magento Admin

This is exactly what this issue is about; If you create a new type from the backend or through a setup script, you cannot reference it in the view.xml because the default view.xsd validator is used on it to validate if your type is one of the allowed values. The validator will say the new type is invalid because it does not know about it.

**Comment by magento-engcom-team** [comment]
@erfanimani, thank you for your report.
The issue is already fixed in 2.2.0
